### ISSUE TITLE
Oauth2 Module authorization code flow possible bugs

### ISSUE THREAD

**Issue by psp589** [OrchardCMS/Orchard2#262]
I'm trying to work with the oauth authorization code flow, but I was always receiving an error with postman "Could not complete OAuth2.0 login" until I've removed the ValidateAntiForgeryToken tag from the AccessController class :

![image](https://cloud.githubusercontent.com/assets/5736566/19074217/7c63e72c-8a3f-11e6-93fc-498c55bd7fd8.png)

After that the login window pops up and then the window where you allow or deny the access to your app. But I can't get anything else. Activating the debug level in the openid module statup.cs I see that
the method Accept (in AccesController) has the request as null ( result of HttpContext.GetOpenIdConnectRequest()), sending it to the method IssueAccessIdentityTokens will throw an exception here :

![image](https://cloud.githubusercontent.com/assets/5736566/19074612/46c55978-8a41-11e6-9773-b1fe3a74e69d.png)

Could anyone verify if I'm doing something wrong or there is some bug over there? 

Thank you.

**Comment by PinpointTownes** [comment]
Thanks for reporting this bug.

> I'm trying to work with the oauth authorization code flow, but I was always receiving an error with postman "Could not complete OAuth2.0 login" until I've removed the ValidateAntiForgeryToken tag from the AccessController class :

It's definitely not how this should work, as antiforgery must not disabled for an interactive endpoint like the authorization endpoint. 

> Could anyone verify if I'm doing something wrong or there is some bug over there?

I'll take a look tonight, but I'm fairly confident that's because `Accept` doesn't have an appropriate route template. For OpenIddict to treat the incoming request as an authorization request, the request path must begin with the route defined when calling `EnableAuthorizationEndpoint`.

I suppose we should also add a bunch of if or `Debug.Assert` to ensure we never end up with an empty `request` (which is something the OpenIddict MVC binder does, but Orchard doesn't use it).

**Comment by psp589** [comment]
I don't fully understand how it works, but  what you mean is that OpenIddictBuild needs another method 
"EnableAcceptEndpoint" where we can stablish the Accept route?

**Comment by PinpointTownes** [comment]
FYI, I sent a PR that should fix all the bugs I encountered when debugging this module: https://github.com/OrchardCMS/Orchard2/pull/264.

> I don't fully understand how it works, but what you mean is that OpenIddictBuild needs another method 
"EnableAcceptEndpoint" where we can stablish the Accept route?

Actually, I opted for a different approach: `Authorize()`/`Accept()`/`Deny()` will now share the same action name and the same route (`[FormValueRequired]` is used to discriminate them).
